csharp/src/Ice/IncomingResponseFrame.cs

```diff
@@ -13,105 +13,155 @@ public enum ResultType : byte { Success, Failure };
+        /// <summary>Reads an empty return value from the response frame.</summary>
+        public void ReadVoidReturnValue()
         {
-            InputStream.StartEncapsulation();
             switch (ReplyStatus)
             {
                 case ReplyStatus.OK:
-                    return (ResultType.Success, InputStream);
+                {
+                    var istr = new InputStream(_communicator, Payload, 1);
+                    istr.StartEncapsulation();
+                    istr.EndEncapsulation();
+                    break;
+                }
                 case ReplyStatus.UserException:
-                    return (ResultType.Failure, InputStream);
-                default:
-                    throw new InvalidOperationException(); // TODO: better exception and message
+                {
+                    var istr = new InputStream(_communicator, Payload, 1);
+                    istr.StartEncapsulation();
+                    RemoteException ex = istr.ReadException();
+                    istr.EndEncapsulation();
+                    throw ex;
+                }
+                case ReplyStatus.ObjectNotExistException:
+                case ReplyStatus.FacetNotExistException:
+                case ReplyStatus.OperationNotExistException:
+                {
+                    throw ReadDispatchException();
+                }
+                case ReplyStatus.UnknownException:
+                case ReplyStatus.UnknownLocalException:
+                case ReplyStatus.UnknownUserException:
+                {
+                    throw ReadUnhandledException();
+                }
             }
         }
 
-        /// <summary>Reads the return value carried by this response frame using the provided InputStreamReader.
-        /// If the response frame carries a failure, ReadReturnValue reads and throws this exception.</summary>
-        /// <param name="reader">InputStreamReader that reads the return value.</param>
-        /// <returns>The return value.</returns>
-        public T ReadReturnValue<T>(InputStreamReader<T> reader)
+        public IncomingResponseFrame(Communicator communicator, ArraySegment<byte> payload)
         {
-            if (ReadResult().ResultType == ResultType.Failure)
+            _communicator = communicator;
+            byte replyStatus = payload[0];
+            if (replyStatus < 0 || replyStatus > 7)
             {
-                // TODO: would be nicer to read then EndEncaps then throw the exception
-                InputStream.ThrowException();
+                throw new InvalidDataException(
+                    $"received ice1 response frame with unknown reply status `{replyStatus}'");
+            }
+            ReplyStatus = (ReplyStatus)replyStatus;
+            Payload = payload;
+            if (ReplyStatus == ReplyStatus.UserException || ReplyStatus == ReplyStatus.OK)
+            {
+                Encoding = new Encoding(payload[5], payload[6]);
```

Comment:
I would make sense to read and check the size of the encaps, not just skip it to read the encoding!

Reply:
I added a check for the size